sdk docs (#734)
flowglad/flowglad: platform/docs/sdks/react.mdx

---
title: "@flowglad/react"
description: "React components and hooks for Flowglad billing and subscription management"
---

import ReactSetup from '/snippets/setup-react.mdx'

<Info>
  NPM Package: [@flowglad/react](https://www.npmjs.com/package/@flowglad/react)
</Info>

## Overview

The React SDK provides components, hooks, and context for managing billing and subscriptions in your React applications. It works with any React framework and requires a backend server running the Flowglad Server SDK.

<ReactSetup />

## Key Features

- **Type-safe Hooks**: Access billing data with full TypeScript support
- **React Context**: Global state management for billing information
- **Feature Access Control**: Check user access to features
- **Checkout Sessions**: Create and manage payment flows
- **Subscription Management**: Handle subscription lifecycles

## API Reference

### FlowgladProvider

The main provider component that wraps your application.

#### Props

```typescript
interface FlowgladProviderProps {

>   loadBilling?: boolean

Reviewer: +1

  
  // Loading states
  isLoading: boolean
  error: Error | null
  
  // Refresh data
  refetch: () => Promise<void>
  reload: () => Promise<void>
  loadBilling: boolean
  loaded: boolean
  errors: Error[] | null
}
```

#### Example Usage

```tsx
import { useBilling } from '@flowglad/react'

function FeatureGate({ featureId, children }) {
  const { checkFeatureAccess, createCheckoutSession } = useBilling()

  if (!checkFeatureAccess) {
    return <div>Loading...</div>
  }

  if (!checkFeatureAccess(featureId)) {
    return (
      <div>
        <p>Upgrade to access this feature</p>
        <button
          onClick={() =>
            createCheckoutSession({
              priceId: 'price_123',
              successUrl: window.location.href,
              cancelUrl: window.location.href,
              autoRedirect: true,
            })
          }
        >
          Upgrade Now
        </button>
      </div>
    )
  }

  return <>{children}</>
}
```

### createCheckoutSession

Create a new checkout session for subscriptions or one-time payments.

#### Parameters

```typescript
interface CheckoutSessionParams {

>   priceId: string

Reviewer: ```suggestion
  priceId?: string
  priceSlug?: string
```

  successUrl: string
  cancelUrl: string
  autoRedirect?: boolean
  quantity?: number
  outputMetadata?: Record<string, unknown>
  outputName?: string
}
```

#### Example

```tsx
const handleUpgrade = async () => {
  await createCheckoutSession({
    priceId: 'price_premium_monthly',
    successUrl: `${window.location.origin}/billing/success`,
    cancelUrl: `${window.location.origin}/billing`,
    autoRedirect: true,
    quantity: 1,
  })
}
```

### createAddPaymentMethodCheckoutSession

Open a Flowglad-hosted flow that collects and stores a new payment method for the signed-in customer.

#### Parameters

```typescript
interface AddPaymentMethodCheckoutParams {
  successUrl: string
  cancelUrl: string
  autoRedirect?: boolean
  targetSubscriptionId?: string
}
```

#### Example

```tsx
await createAddPaymentMethodCheckoutSession({
  successUrl: `${window.location.origin}/billing/payment-methods`,
  cancelUrl: window.location.href,
  autoRedirect: true,
})
```

### createActivateSubscriptionCheckoutSession

> Trigger the activation flow for an existing subscription that needs a payment method before billing can start.

Reviewer: ```suggestion
Trigger the activation flow for an existing subscription that needs a payment method before billing can start. Activations can happen for subscriptions on trial, or subscriptions where payment is due but has not yet been completed.
```

#### Parameters

```typescript
interface ActivateSubscriptionCheckoutParams {

>   targetSubscriptionId: string

Reviewer: ```suggestion
  /*
   * The subscription to activate
   */
  targetSubscriptionId: string
```

  priceId: string
  successUrl: string
  cancelUrl: string
  autoRedirect?: boolean
}
```

#### Example

```tsx
await createActivateSubscriptionCheckoutSession({
  targetSubscriptionId: 'sub_123',
  priceId: 'price_premium_monthly',
  successUrl: `${window.location.origin}/billing/success`,
  cancelUrl: window.location.href,
  autoRedirect: true,
})
```

### cancelSubscription

Cancel a subscription owned by the current customer and refresh billing data.

#### Parameters

```typescript
interface CancelSubscriptionParams {
  id: string
  cancellation:
    | { timing: 'at_end_of_current_billing_period' }
    | { timing: 'at_future_date'; endDate: Date }
    | { timing: 'immediately' }
}
```

#### Example

```tsx
await cancelSubscription({
  id: 'sub_123',
  cancellation: {
    timing: 'at_end_of_current_billing_period',
  },
})
```

### checkFeatureAccess

Check if the current customer has access to a specific feature.

#### Parameters

> - `featureId: string` - The ID of the feature to check

Reviewer: ```suggestion
- `featureSlug: string` - The slug of the feature to check
```

#### Returns

- `boolean` - `true` if the customer has access, `false` otherwise

#### Example

```tsx
const hasAdvancedAnalytics = checkFeatureAccess('advanced_analytics')

if (hasAdvancedAnalytics) {
  return <AdvancedAnalytics />
}

return <BasicAnalytics />
```

### checkUsageBalance

Get the remaining balance for a usage meter tied to the customer’s subscriptions.

#### Parameters

- `usageMeterSlug: string` – Slug of the usage meter to inspect.
- `options.subscriptionId?: string` – Limit the check to a specific subscription (optional).

#### Returns

- `{ availableBalance: number } | null`

#### Example

```tsx
const usage = checkUsageBalance('api_calls')
return usage ? (
  <p>{usage.availableBalance} calls remaining</p>
) : (
  <p>No usage meter found.</p>
)
```

### getProduct

Look up a specific product from the catalog that ships with billing data.

#### Parameters

- `productSlug: string`

#### Returns

> - `CatalogProduct | null`

Reviewer: ```suggestion
- `Product | null`
```

    {monthly.intervalUnit ? ` / ${monthly.intervalUnit}` : ''}
  </span>
) : (
  <span>Price not available</span>
)
```

### Other billing records

The billing payload also exposes higher-level billing records you can use to build richer UI:

- `customer` – The Flowglad customer record (including email, name, IDs).
- `subscriptions` – All subscriptions (active and past) returned for the customer.
- `purchases` – Historical purchases for the customer.
- `invoices` – Invoice records (including status and totals).
- `currentSubscriptions` – Subscriptions currently active or pending.
- `paymentMethods` – Saved payment methods for the customer.
- `billingPortalUrl` – Deep link to Flowglad’s hosted billing portal (if enabled).
- `pricingModel` – Resolved pricing model associated with the customer’s plan.

```tsx
const { customer, subscriptions, paymentMethods, billingPortalUrl } = useBilling()

const activeSub = subscriptions?.find((sub) => sub.status === 'active')
const defaultMethod = paymentMethods?.[0]
```

## Common Patterns

### Feature Gating Component

```tsx
import { useBilling } from '@flowglad/react'

export function FeatureGate({ 
  featureId, 
  fallback, 
  children 
}: {

>   featureId: string

Reviewer: ```suggestion
  featureSlug, 
  fallback, 
  children 
}: {
  featureSlug: string
```

  fallback?: React.ReactNode
  children: React.ReactNode
}) {
  const { checkFeatureAccess } = useBilling()

  if (!checkFeatureAccess) {
    return null
  }

  if (!checkFeatureAccess(featureId)) {
    return fallback || <div>Access denied</div>
  }

  return <>{children}</>
}

// Usage
<FeatureGate 

>   featureId="premium_feature"

Reviewer: ```suggestion
  featureSlug="premium_feature"
```

Create a new checkout session for subscriptions or one-time payments.

Checkout sessions for specific prices can be made using either `priceSlug` (an identifier that you define), or `priceId` (an identifier Flowglad defines) as a parameter. 

`priceSlug` is recommended, so that you don't need to store any references to Flowglad ids in your application database or environment variables.

#### Parameters

```typescript
interface CheckoutSessionParams {

>   // Must specify either priceId or priceSlug

Reviewer: ```suggestion
  /*
   * Must specify either priceId or priceSlug
   */
```
